### PR TITLE
fix: retain user-entered multiProduct info

### PR DESCRIPTION
**Pull request by dannydaviesIW**
# Description

-   This PR refactors the MULTIPLE_PRODUCT_ATTRIBUTE so that the information the user has entered on the /multipleProducts page is retained, even when they have progressed further through the journey.
 
# Testing instructions

-   Pull down this branch and run the site locally. Navigate through a multi product period ticket journey (either of geoZone or multiServices) and check that the multipleProducts, multipleProductValidity and ticketConfirmation pages behave as they did previously.
-   Once you have navigated through the journey to the ticketConfirmation page, click one of the 'Change' links that will take you back to the /multipleProducts page and check that the info you previously entered is rendered on the page.

# Type of change

-   [ ] feat - A new feature
-   [X] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

# Checklist:

-   [X] Able to run pr locally
-   [X] Followed acceptance criteria
-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my own code
-   [ ] I have made corresponding changes to the documentation if applicable
-   [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tests/pages/api/multipleProductValidity.test.ts [hunk 1]
```diff
@@ -1,169 +1,68 @@
...
-                productDuration: '30',
-                productDurationId: '',
-                productValidity: 'endOfServiceDay',
-            };
-            const result = addErrorsIfInvalid(req, product, userInputIndex);
-
-            expect(result.productValidity).toBe('endOfServiceDay');
-            expect(result.productValidityError).toBe(undefined);
+        expect(writeHeadMock).toBeCalledWith(302, {
+            Location: '/ticketConfirmation',
```
Comment (matt-twomey) [inline tests/pages/api/multipleProductValidity.test.ts:39]: Why is this location different to the redirect expected according to this test's 'it' description?

File: tests/testData/mockData.ts [hunk 2]
```diff
@@ -2082,6 +2109,47 @@ export const invalidDurationProducts: MultiProduct[] = [
...
+        productName: 'valid duration',
+        productNameId: 'multiple-product-name-input-0',
+        productPrice: '.',
+        productPriceId: '.',
+        productDuration: '66',
+        productDurationId: '.',
+        productDurationUnits: 'week',
+        productDurationUnitsId: 'product-duration-units-1',
+        productValidity: '',
+        productValidityId: '',
```
Comment (matt-twomey) [inline tests/testData/mockData.ts:2152]: This appears to be identical to the `valid duration type` product - is it necessary?

File: tests/pages/api/multipleProducts.test.ts [hunk 3]
```diff
@@ -170,4 +123,75 @@ describe('multipleProduct', () => {
...
+        });
+    });
+
+    describe('checkProductDurationTypesAreValid', () => {
+        it('adds duration errors to a product with invalid durations', () => {
+            const result = checkProductDurationTypesAreValid(invalidDurationTypeProducts);
+            expect(result[0].productDurationUnitsError).toBeUndefined();
+            expect(result[1].productDurationUnitsError).toBe('Choose an option from the dropdown');
+            expect(result[2].productDurationUnitsError).toBeUndefined();
+        });
```
Comment (matt-twomey) [inline tests/pages/api/multipleProducts.test.ts:168]: This third test appears to test the same thing as that on line 165 (the `valid duration type` and `valid duration` products in the `invalidDurationTypeProducts` mock appear to be identical?)

File: src/pages/api/multipleProductValidity.ts [hunk 4]
```diff
@@ -93,19 +18,53 @@ export default (req: NextApiRequestWithSession, res: NextApiResponse): void => {
...
+            },
+        );
 
-        updateSessionAttribute(req, MULTIPLE_PRODUCT_ATTRIBUTE, {
-            products,
-            endTimesList: req.body && req.body.listOfEndTimes !== '' ? req.body.listOfEndTimes.split(',') : [],
+        const errors: ErrorInfo[] = [];
+        products.forEach(product => {
+            const index = product.productValidityId.split('-').pop();
+            const endTimeId = `validity-end-time-${index}`;
```
[the remark this answers is not in the document]
Comment (RobCatton) [reply src/pages/api/multipleProductValidity.ts:40]: I dont think so, as its an optional thing on a product, so on its first pass through the API, I don't think they would have id's.

File: src/pages/api/multipleProducts.ts [hunk 5]
```diff
@@ -194,14 +199,16 @@ export default (req: NextApiRequestWithSession, res: NextApiResponse): void => {
...
+                productValidityId: '',
             };
             multipleProducts.push(product);
         }
 
-        const nameValidationResult: MultiProduct[] = checkProductNamesAreValid(multipleProducts);
-        const priceValidationResult: MultiProduct[] = checkProductPricesAreValid(nameValidationResult);
-        const productDurationResult: MultiProduct[] = checkProductDurationsAreValid(priceValidationResult);
-        const fullValidationResult: MultiProduct[] = checkProductDurationTypesAreValid(productDurationResult);
+        const nameValidationResult = checkProductNamesAreValid(multipleProducts);
```
[the remark this answers is not in the document]
Comment (RobCatton) [reply src/pages/api/multipleProducts.ts:208]: Readded.

File: src/pages/multipleProductValidity.tsx [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable jsx-a11y/no-onchange */
```
[the remark this answers is not in the document]
Comment (RobCatton) [reply src/pages/multipleProductValidity.tsx:1]: Yeah, this comment tells the linter to ignore something we're doing and let us do it - in this instance, using an on-change.

File: tests/pages/api/multipleProductValidity.test.ts (hunk 1 above)
Comment (RobCatton) [reply tests/pages/api/multipleProductValidity.test.ts:39]: good catch

**Review by RobCatton — APPROVED** [review]

**Pull request merged**